nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,565 @@
+
+```javascript
+[
+    body("name")
+    .trim()
+    .isLength({ min: 1 })
+    .withMessage("Name can not be empty.")
+    .isAlpha()
+    .withMessage("Name must only contain alphabet letters."),  
+];
+```
+
+This ensures that `name` is not only present and trimmed, but also only contains alphabet letters.
+
+While this might work for outputs we know won't have special characters, like names or ages, we also have to consider situations that do allow those characters. For example, when writing their "About Me" description, what would happen if the client decides to inject JavaScript code instead?
+
+```ejs
+<div>
+  About Me: <%- description %>
+</div>
+
+// The client then inputs the following as their page's About Me:
+<script>alert("Hacked!");</script>
+```
+
+When unescaped, this would be rendered into HTML as:
+
+```html
+<div>
+  About Me: <script>alert("Hacked!");</script>!
+</div>
+```
+
+To prevent this [cross-site scripting (XSS) attack](https://en.wikipedia.org/wiki/Cross-site_scripting), we can *escape* the output (you may also see this referred to as *encoding*). Escaped HTML replaces special characters, like `<`, with their respective HTML entities, in this case `&lt;`. In EJS, we can escape the output using `<%= %>`.
+
+```ejs
+<div>
+  About Me: <%= username %>!
+</div>
+
+// The escaped output is now rendered harmless:
+// About Me: &lt;script&gt;alert(&quot;Hacked!&quot;);&lt;/script&gt;!
+```
+
+So why not escape the data when we receive it by adding .escape() to the end of our body() validation chain instead? "Dangerous characters" are only dangerous at the point of use, and also only in certain contexts. What's "dangerous" for HTML may not be dangerous for SQL and vice versa, and they won't pose risks until they get used in those contexts.
+
+Also, if we have data with HTML entities after escaping them, if we used escaped output with (<%= %>), then &lt; wouldn't become < but output literally as the text &lt;. We'd need to either unescape them then re-escape them via escaped output, or end up using unescaped output (<%- %>), which, like using .innerHTML, is undesirable as shown above.
+
+### Validation results
+
+Once the validation rules are applied, you can use 'validationResult' to handle any validation errors. We use `asyncHandler` to automatically catch these errors in our async route handlers and pass them to the middleware:
+
+```javascript
+// asyncHandler lets us wrap async express routes to handle errors.
+asyncHandler(async (req, res, next) => {
+  const errors = validationResult(req);
+  
+  if (!errors.isEmpty()) {
+    return res.status(400).json({ errors: errors.array() });
+  } else {
+    res.send("Data is valid, form submitted!");
+  }
+});
+```
+
+This setup checks for any failed validation checks, and if there are any (the errors array is NOT empty), then it sends a 400 code, along with any errors that may be present.
```

Comment:
Need to change to use `res.render` to render the appropriate page (e.g. the form page but with the errors listed, or if successful, perhaps a redirect to the homepage or wherever is most appropriate).
Sending JSON or a string would be more relevant for APIs later but not when we're still using view templates.

The last paragraph will also need adjusting to reflect the change.